docs/booster.md

```diff
@@ -0,0 +1,231 @@
+                                                                    ▼                         ▼
+
+                                                           Return rewritten          Return all branches
+```
+
+The rule application procedure trys applying rules, grouped by priorities. If one or more rules from the current priority group apply, the rewritten configuration(s) get(s) returned. Otherwise, lower priority rules are tried until potentially all groups have been tried. If no rules applied, a stuck result is returned. 
+
+The process of applying a single rule involves several stages. First, the unification procedure tries to match the LHS of a rewrite rule with the configuration. The unification procedure is more limited than in the old backend and a common source of failed unification is AC unification. Specifically, the booster can currently only perform unification of a small subset of restricted cases for maps, namely unifying concrete keys in the rule with concrete keys in the configuration. In practice, this covers a good portion of rewrite rules encountered in the real world. If unification fails with a complicated AC problem, it currently falls back onto the old backend.
+
+If unification succeeds, the resulting substitution is checked to be a "matching" substitution, i.e. the variables of the substitution map are precisely only those that appear in the LHS of the rule. This substitution is then applied to the rule's RHS, which becomes the new configuration. Before a sucessful rewrite can be returned, the "requires" clauses/predicates of the rule have to be discharged. This means that all the predicates of the rule have to evaluate to true if the rule application is to be succesful, or at least one evaluates to False if the rule does not apply. Predicate simplification proceeds in much the same way as the already mentioned configuration wide simplification step. However instead of applying function and simplification rules to the configuration, we instead call the simplifier on each of the *requires* clauses of the rule. If any of the predicates cannot be simplified to true/false, the booster proceeds to collect all such unknow predicates and uses an SMT solver to try to either derive a contradiction from the set of unknown predicates, in which case the rule aplication fails, or otherwise derive true. If the result is unknown again, the booster falls back onto the old backend.
+
+If all the requires clauses have been successfully proven, the booster proceeds to check all the nesures clauses individually, in case any evaluate to false. It then calls the SMT solver again with the full set of current configuration constraints, also adding the new ensures clauses. If this new set of constraints is false/`_|_`, then the whole configuration is `_|_` and the booster records this rewrite as being trivial.
+
+### Simplification / function evaluation
```

Comment:
After reading this section, I'm left asking for more. I want this section to flow similarly to the rewriting section:
* we describe how we select which equations to attempt
* we describe in detail the process of application of a single equation, perhaps with a diagram

What I also want to convey is this section is the encoding of function rules and simplifications.